### PR TITLE
Florence test with light forcing & validation data

### PR DESCRIPTION
**Pull request by awlostowski-noaa**
This pull request introduces a new test to the repository. The new test considers the Hurricane Florence event on the Falls Lake basin, near Durham North Carolina. While forcing data (lateral inflows) and validation data are from a WRF Hydro simulation, all input models are provided as flat files derived from a bulkier set of CHRTOUT files. Routing model parameters for the simulation domain are specified in `test/input/yaml/Florence_Benchmark.yaml`.

## New files added
- `test/input/florence_933020089/RESTART/HYDRO_RST.2018-09-XX_00_00_DOMAIN1`
- `test/input/florence_933020089/RESTART/ChannelRestart_2018-09-XX.csv`
- `test/input/florence_933020089/OUTPUT/SeptemberFlorenceStreamflow_wDates.csv`
- `test/input/florence_933020089/OUTPUT/SeptemberFlorenceQlateral_wDates.csv`
- `test/input/florence_933020089/OUTPUT/SeptemberFlorenceQlateral.csv`
- `test/input/florence_933020089/DOMAIN/Route_Link.nc`
- `test/input/florence_933020089/DOMAIN/LAKEPARM.nc`
- `test/input/geo/Channels/masks/933020089_mask.csv`
- `test/input/yaml/Florence_Benchmark.yaml`
-  `test/input/florence_933020089/README.md`

## Files changed to accommodate test capability
- `src/python_routing_v02/build_tests.py`
- `src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py`

### CODE REVIEW

**Comment by awlostowski-noaa** [comment]
TO DO

- [ ] Parity test. Will require reading WRF flow from .csv, rather than CHRTOUTs
- [ ] Write readme file describing the genesis of the WRF-Hydro data
- [ ] Enable capability to read forcing data from .csv with dates as column headers. Currently csv forcing read only works for files with integer timestep headers.

**Comment by awlostowski-noaa** [comment]
Planning to adjust `build_tests.parity_check` to allows validation data to be read from either a list of CHRTOUT files of a single csv file. 

```python
    if parity_parameters["parity_check_input_folder"]:
        # Glob list of CHRTOUT files for validation
        validation_files = glob.glob(
            parity_parameters["parity_check_input_folder"]
            + parity_parameters["parity_check_file_pattern_filter"],
            recursive=True,
        )

        compare_node = parity_parameters["parity_check_compare_node"]

        # read validation data from CHRTOUT files
        validation_data = nhd_io.get_ql_from_wrf_hydro_mf(
            validation_files,
            parity_parameters["parity_check_file_index_col"],
            parity_parameters["parity_check_file_value_col"],
            # [compare_node],
        )
    
    if parity_parameters["parity_check_file"]:
        validatiion_data = pd.csv_read(parity_parameters["parity_check_file"], index_col=0)
        validatiion_data.index = validatiion_data.index.astype(int)
        validatiion_data = validatiion_data.sort_index(axis="index")
```

File: src/python_routing_v02/build_tests.py [hunk 1]
```diff
@@ -161,22 +161,31 @@ def build_test_parameters(
...
+            # [compare_node],
+        )
+    
+    if "parity_check_file" in parity_parameters.keys():
+        validation_data = pd.read_csv(parity_parameters["parity_check_file"], index_col=0)
+        validation_data.index = validation_data.index.astype(int)
+        validation_data.columns = validation_data.columns.astype("datetime64[ns]")
+        validation_data = validation_data.sort_index(axis="index")
+        
     compare_node = parity_parameters["parity_check_compare_node"]
```
Comment (awlostowski-noaa) [inline src/python_routing_v02/build_tests.py:187]: Originally, `validation_data` was constructed from a series of CHRTOUT files stored in `parity_parameters["parity_check_input_folder"]`. I added an option to read validation data straight from a .csv file. This happens if `parity_parameters["parity_check_file"]` is provided.

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py [hunk 2]
```diff
@@ -867,7 +867,7 @@ def main():
     if showtiming:
         print("... in %s seconds." % (time.time() - start_time))
 
-    if "parity_check_input_folder" in parity_parameters:
+    if "parity_check_input_folder" in parity_parameters or "parity_check_file" in parity_parameters:
```
Comment (awlostowski-noaa) [inline src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:870]: To allow for parity assessment based on validation data stored in .csv, I appended this if statement to work if either "parity_check_input_folder" or "parity_check_file" is in paraity_parameters

File: test/input/yaml/Florence_Benchmark.yaml [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    #WRF-Hydro waterbodies restart file
+    wrf_hydro_waterbody_restart_file: "../../test/input/florence_933020089/RESTART/HYDRO_RST.2018-09-01_00_00_DOMAIN1"
+    #WRF-Hydro waterbody ID crosswalk file
+    wrf_hydro_waterbody_ID_crosswalk_file: "../../test/input/florence_933020089/DOMAIN/LAKEPARM.nc"
+    wrf_hydro_waterbody_ID_crosswalk_file_field_name: lake_id
+    #WRF-Hydro waterbody crosswalk filter file
+    wrf_hydro_waterbody_crosswalk_filter_file: "../../test/input/florence_933020089/DOMAIN/Route_Link.nc"
+    wrf_hydro_waterbody_crosswalk_filter_file_field_name: NHDWaterbodyComID
+parity_parameters:
+    parity_check_file: "../../test/input/florence_933020089/OUTPUT/SeptemberFlorenceStreamflow_wDates.csv"
```
Comment (awlostowski-noaa) [inline test/input/yaml/Florence_Benchmark.yaml:101]: Notice we are comparing against data in a .csv file (light), not observations within a series of CHRTOUT files (heavy)

**Comment by awlostowski-noaa** [comment]
To test, I ran the Florence simulation from yaml input. The command line output below shows parity.

```
$ python compute_nhd_routing_SingleSeg_v02.py -f ../../test/input/yaml/Florence_Benchmark.yaml

creating supernetwork connections set
supernetwork connections set complete
... in 0.08495378494262695 seconds.
organizing connections into reaches ...
reach organization complete
... in 0.0022084712982177734 seconds.
setting channel initial states ...
channel initial states complete
... in 0.061678171157836914 seconds.
creating qlateral array ...
qlateral array complete
... in 0.18738818168640137 seconds.
executing routing computation ...
JIT Preprocessing time 0.0035109519958496094 seconds.
starting Parallel JIT calculation
PARALLEL TIME 10.442301273345947 seconds.
           (0, q)    (0, v)    (0, d)    (1, q)    (1, v)    (1, d)    (2, q)    (2, v)    (2, d)    (3, q)  ...  (8636, d)  (8637, q)  (8637, v)  (8637, d)  (8638, q)  (8638, v)  (8638, d)  (8639, q)  (8639, v)  (8639, d)
8777177  0.001673  0.087550  0.011367  0.001672  0.087548  0.011367  0.001672  0.087547  0.011366  0.001672  ...   0.022418   0.005078   0.136482   0.022418   0.005078   0.136481   0.022418   0.005078   0.136479   0.022417
8777181  0.000115  0.032804  0.001948  0.000115  0.032804  0.001948  0.000115  0.032804  0.001948  0.000115  ...   0.005523   0.000498   0.065437   0.005523   0.000498   0.065436   0.005523   0.000498   0.065436   0.005523
8777199  0.006617  0.093254  0.026914  0.006615  0.093246  0.026910  0.006614  0.093238  0.026906  0.006612  ...   0.086166   0.045931   0.196692   0.086284   0.046033   0.196854   0.086398   0.046131   0.197009   0.086507
8777201  0.002673  0.097234  0.013301  0.002672  0.097230  0.013300  0.002672  0.097227  0.013300  0.002672  ...   0.024332   0.007197   0.144369   0.024332   0.007197   0.144369   0.024332   0.007197   0.144369   0.024332
8777269  0.001909  0.074826  0.009602  0.001909  0.074821  0.009601  0.001909  0.074817  0.009600  0.001908  ...   0.034513   0.013317   0.172869   0.034511   0.013316   0.172864   0.034510   0.013315   0.172860   0.034509
...           ...       ...       ...       ...       ...       ...       ...       ...       ...       ...  ...        ...        ...        ...        ...        ...        ...        ...        ...        ...        ...
8776671  0.009901  0.229097  0.034547  0.009901  0.229100  0.034548  0.009901  0.229102  0.034548  0.009902  ...   0.043222   0.014348   0.263848   0.043220   0.014347   0.263843   0.043219   0.014347   0.263839   0.043218
8776679  0.022819  0.158373  0.061525  0.022821  0.158379  0.061529  0.022823  0.158384  0.061532  0.022825  ...   0.147980   0.099386   0.272065   0.147970   0.099376   0.272056   0.147961   0.099367   0.272048   0.147954
8778589  0.001842  0.011918  0.129591  0.001842  0.011918  0.129589  0.001842  0.011918  0.129586  0.001842  ...   0.273732   0.006560   0.021809   0.401306   0.006559   0.017904   0.273683   0.006559   0.021808   0.401267
8778991  0.000511  0.059265  0.005908  0.000511  0.059265  0.005908  0.000511  0.059265  0.005908  0.000511  ...   0.024477   0.004202   0.149836   0.024477   0.004202   0.149836   0.024477   0.004202   0.149836   0.024477
8778995  0.000046  0.003755  0.020157  0.000046  0.003755  0.020156  0.000046  0.003755  0.020155  0.000046  ...   0.121994   0.000908   0.010791   0.121981   0.000908   0.010790   0.121968   0.000908   0.010790   0.121956

[1107 rows x 25920 columns]
ordered reach computation complete
... in 10.9276762008667 seconds.
conducting parity check, comparing WRF Hydro results against t-route results
                     flow, wrf (cms)  flow, t-route (cms)
2018-09-01 00:00:00         0.000549             0.000550
2018-09-01 01:00:00         0.000551             0.000551
2018-09-01 02:00:00         0.000551             0.000551
2018-09-01 03:00:00         0.000551             0.000551
2018-09-01 04:00:00         0.000572             0.000572
...                              ...                  ...
2018-09-30 19:00:00         0.003900             0.003900
2018-09-30 20:00:00         0.004174             0.004174
2018-09-30 21:00:00         0.004482             0.004482
2018-09-30 22:00:00         0.004556             0.004556
2018-09-30 23:00:00         0.004353             0.004353
```

**Comment by awlostowski-noaa** [comment]
@awlostowski-noaa 

- [ ] Create readme
- [ ] Duplicate forcing and validation files?
- [ ] Read forcing data csv with dates for column headers

File: src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py (hunk 2 above)
Comment (jameshalgren) [reply src/python_routing_v02/compute_nhd_routing_SingleSeg_v02.py:870]: Might consider a check on the length of parity_parameters...

File: src/python_routing_v02/build_tests.py (hunk 1 above)
Comment (jameshalgren) [reply src/python_routing_v02/build_tests.py]: @awlostowski-noaa
The code above shows the logic you describe, but it looks like it was lost in a later commit... I'm working on restoring it here for a merge candidate.

**Comment by awlostowski-noaa** [comment]
ping @jameshalgren mistakes in `build_tests.py` fixed. Custom Input yaml parameterized test case worked fine for me.

**Comment by jameshalgren** [comment]
ping @jhrehanoaa 
Let's try to finish without
`test/input/florence_933020089/RESTART/HYDRO_RST.2018-09-XX_00_00_DOMAIN1`
and
`test/input/florence_933020089/DOMAIN/LAKEPARM.nc`

Then this should be clean. 

- [ ] Make a new issue to add a reservoir restart and reservoir file to this test (will require permission for LAKEPARM)

**Comment by jameshalgren** [comment]
- [ ] Evaluate impact of additional output (Cn Ck X) on CONUS execution...

**Comment by jameshalgren** [comment]
- [ ] TODO: Make new issue regarding CSV file filtering to apply currently-unused options in .yaml control file.

File: test/input/yaml/Florence_Benchmark.yaml (hunk 3 above)
Comment (jameshalgren) [reply test/input/yaml/Florence_Benchmark.yaml:105]: 10-4. Good.

**Comment by jameshalgren** [comment]
> * [x]  Evaluate impact of additional output (Cn Ck X) on CONUS execution...

Executed CONUS benchmark -- timing is unaffected. Size of return causes bus error for a 5 day simulation, 1440 time steps on Cheyenne. 864 timesteps fits within the memory limits and executes without problem. This is the case __even if the courant output is not selected!__

**Comment by jameshalgren** [comment]
Testcases: 
- [x] `python compute_nhd_routing_SingleSeg_v02.py --test pocono1 -t -v --debuglevel 1`
- [x] `python compute_nhd_routing_SingleSeg_v02.py -f ../../test/input/yaml/CustomInput.yaml`
- [x] `python compute_nhd_routing_SingleSeg_v02.py -t -n Mainstems_CONUS -v --debuglevel 1 --parallel by-subnetwork-jit-clustered --subnet-size 100 --nts 288 --cpu-pool  6`
- [x] ` python compute_nhd_routing_SingleSeg_v02.py -f ../../test/input/yaml/Florence_Benchmark.yaml`

**Review by jameshalgren — APPROVED** [review]
Have created issues to address any remaining concerns.

**Pull request merged**